app/main.py

```diff
@@ -1,214 +1,257 @@
 
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
+        # apply potion if exist
+        if red_knight["potion"] is not None:
+            if "power" in red_knight["potion"]["effect"]:
+                red_knight["power"] += red_knight["potion"]["effect"]["power"]
 
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
+            if "protection" in red_knight["potion"]["effect"]:
+                red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
 
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
+            if "hp" in red_knight["potion"]["effect"]:
+                red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
 
-    # -------------------------------------------------------------------------------
-    # BATTLE:
+        # -------------------------------------------------------------------------------
+        # BATTLE:
 
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
+        # 1 Lancelot vs Mordred:
+        lancelot["hp"] -= mordred["power"] - lancelot["protection"]
+        mordred["hp"] -= lancelot["power"] - mordred["protection"]
 
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
+        # check if someone fell in battle
+        if lancelot["hp"] <= 0:
+            lancelot["hp"] = 0
 
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
+        if mordred["hp"] <= 0:
+            mordred["hp"] = 0
 
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
+        # 2 Arthur vs Red Knight:
+        arthur["hp"] -= red_knight["power"] - arthur["protection"]
+        red_knight["hp"] -= arthur["power"] - red_knight["protection"]
 
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
+        # check if someone fell in battle
+        if arthur["hp"] <= 0:
+            arthur["hp"] = 0
 
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
-
-    # Return battle results:
-    return {
-        lancelot["name"]: lancelot["hp"],
-        arthur["name"]: arthur["hp"],
-        mordred["name"]: mordred["hp"],
-        red_knight["name"]: red_knight["hp"],
-    }
+        if red_knight["hp"] <= 0:
+            red_knight["hp"] = 0
 
+        # Return battle results:
+        return {
+            lancelot["name"]: lancelot["hp"],
+            arthur["name"]: arthur["hp"],
+            mordred["name"]: mordred["hp"],
+            red_knight["name"]: red_knight["hp"],
+        }
```

Comment:
The battle logic in the `battle` function is repetitive and manually calculates the effects of armour, weapons, and potions for each knight. This logic is already encapsulated in the `Knight` class through the `_calculate_stats` method. Consider refactoring this section to create `Knight` instances and use their methods to handle these calculations, which will make the code more concise and maintainable.